internal/auto_heartbeater_test.go

```diff
@@ -0,0 +1,117 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package internal
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap/zaptest"
+
+	"go.uber.org/cadence/.gen/go/shared"
+	"go.uber.org/cadence/internal/common"
+	"go.uber.org/cadence/internal/common/util"
+)
+
+func TestAutoHearbeater_Run(t *testing.T) {
+	var (
+		activityType      = "test-activity"
+		workflowExecution = &shared.WorkflowExecution{
+			WorkflowId: common.StringPtr("test-workflow"),
+			RunId:      common.StringPtr("test-run"),
+		}
+	)
+
+	t.Run("worker stop channel", func(t *testing.T) {
+		stopCh := make(chan struct{})
+		invoker := &MockServiceInvoker{}
+		logger := zaptest.NewLogger(t)
+		clock := clockwork.NewFakeClock()
+		hearbeater := newHeartbeater(stopCh, invoker, logger, clock, activityType, workflowExecution)
+
+		close(stopCh)
+
+		hearbeater.Run(context.Background(), time.Second)
+	})
+	t.Run("context done", func(t *testing.T) {
+		stopCh := make(chan struct{})
+		invoker := &MockServiceInvoker{}
+		logger := zaptest.NewLogger(t)
+		clock := clockwork.NewFakeClock()
+		hearbeater := newHeartbeater(stopCh, invoker, logger, clock, activityType, workflowExecution)
+
+		ctx, cancel := context.WithCancel(context.Background())
+		cancel()
+
+		hearbeater.Run(ctx, time.Second)
+	})
+	t.Run("heartbeat success", func(t *testing.T) {
+		stopCh := make(chan struct{})
+		invoker := &MockServiceInvoker{}
+		invoker.EXPECT().BackgroundHeartbeat().Return(nil).Once()
+		logger := zaptest.NewLogger(t)
+		clock := clockwork.NewFakeClock()
+		hearbeater := newHeartbeater(stopCh, invoker, logger, clock, activityType, workflowExecution)
+
+		ctx, cancel := context.WithCancel(context.Background())
+
+		wg := &sync.WaitGroup{}
+		wg.Add(1)
+
+		go func() {
+			hearbeater.Run(ctx, time.Second)
+			wg.Done()
+		}()
+		clock.BlockUntil(1)
+		clock.Advance(time.Second / 2)
+		cancel()
+
+		require.True(t, util.AwaitWaitGroup(wg, time.Second))
+	})
+	t.Run("heartbeat fail", func(t *testing.T) {
```

Comment:
why not use table test to have less redundancy? this approach is hard to maintain especially while the function under test changes. requires changing N test bodies instead of 1

Reply:
I'm not sure if I can save much with table tests.
I think a setup will have many customizations and be hard to read.